README.md
# date-holidays-be

> Provides the official Belgian holidays in NL/FR/DE/EN.

A zero dependency package to get the official Belgian holidays.

## Install

```sh
npm install --save @itenium/date-holidays-be
```

## Use

```js
import { isHoliday, getHolidays } from '@itenium/date-holidays-be';

const holiday: boolean = isHoliday(new Date(2023, 0, 1));
const holidays: Holiday[] = getHolidays(2023);
```

## Typings

```ts
type Holiday = {
  date: Date;
  labels: {
    nl: string;
    fr: string;
    de: string;
    en: string;
  }
}
```

## Motivation

If you just need to calculate something very specific,
importing [`date-holidays`](https://github.com/commenthol/date-holidays)
results in a whopping 10MB dependency.

This turned out to be a performance issue that
came up during our "React Component Profiling"
technical session at itenium.


## Development

```sh
npm install
npm run build
```


CHANGELOG.md
# Changelog

## 0.0.1 Initial draft
- FEAT: added the functionality to fetch Belgian holidays for a given year
- FEAT: added the functionality to determine whether a given date is a holiday


test/belgianHolidays.test.ts
import {isHoliday, getHolidays} from '../src/belgianHolidays';

describe('Testing the full Belgian holiday dataset', () => {
    test('2023', () => {
        expect(getHolidays(2023)).toEqual([
            {date:new Date(2023, 0, 1), labels: {nl: 'Nieuwjaar', fr: 'Nouvel An', de: 'Neujahr', en: 'New Year\'s Day'}},
            {date:new Date(2023, 3, 10), labels: {nl:'Tweede paasdag', fr: 'Lundi de Pâques', de: 'Ostermontag', en: 'Easter Monday'}},
            {date:new Date(2023, 4, 1), labels: {nl:'Dag van de Arbeid', fr: 'Fête du travail', de: 'Tag der Arbeit', en: 'Labour day'}},
            {date:new Date(2023, 4, 18), labels: {nl:'O.L.H. Hemelvaart', fr: 'Ascension', de: 'Christi Himmelfahrt', en: 'Ascension Day'}},
            {date:new Date(2023, 4, 29), labels: {nl:'Tweede pinksterdag', fr: 'Lundi de Pentecôte', de: 'Pfingstmontag', en: 'Whit Monday'}},
            {date:new Date(2023, 6, 21), labels: {nl: 'Nationale feestdag', fr: 'Fête nationale', de: 'Nationalfeiertag', en: 'National Holiday'}},
            {date:new Date(2023, 7, 15), labels: {nl:'O.L.V. Hemelvaart', fr: 'Assomption', de: 'Mariä Himmelfahrt', en: 'Assumption'}},
            {date:new Date(2023, 10, 1), labels: {nl:'Allerheiligen', fr: 'Toussaint', de: 'Allerheiligen', en: 'All Saints\' Day'}},
            {date:new Date(2023, 10, 11), labels: {nl:'Wapenstilstand', fr:  'Armistice', de: 'Waffenstillstand', en: 'Armistice'}},
            {date:new Date(2023, 11, 25), labels: {nl:'Kerstmis', fr: 'Noël', de: 'Weihnachten', en: 'Christmas Day'}}
        ]);
    });

    test('2020', () => {
        expect(getHolidays(2020)).toEqual([
            {date:new Date(2020, 0, 1), labels: expect.objectContaining({en: 'New Year\'s Day'})},
            {date:new Date(2020, 3, 13), labels: expect.objectContaining({en: 'Easter Monday'})},
            {date:new Date(2020, 4, 1), labels: expect.objectContaining({en: 'Labour day'})},
            {date:new Date(2020, 4, 21), labels: expect.objectContaining({en: 'Ascension Day'})},
            {date:new Date(2020, 5, 1), labels: expect.objectContaining({en: 'Whit Monday'})},
            {date:new Date(2020, 6, 21), labels: expect.objectContaining({en: 'National Holiday'})},
            {date:new Date(2020, 7, 15), labels: expect.objectContaining({en: 'Assumption'})},
            {date:new Date(2020, 10, 1), labels: expect.objectContaining({en: 'All Saints\' Day'})},
            {date:new Date(2020, 10, 11), labels: expect.objectContaining({en: 'Armistice'})},
            {date:new Date(2020, 11, 25), labels: expect.objectContaining({en: 'Christmas Day'})},
        ]);
    });

    test('2019', () => {
        expect(getHolidays(2019)).toEqual([
            {date: new Date(2019, 0, 1), labels: expect.objectContaining({en: 'New Year\'s Day'})},
            {date: new Date(2019, 3, 22), labels: expect.objectContaining({en: 'Easter Monday'})},
            {date: new Date(2019, 4, 1), labels: expect.objectContaining({en: 'Labour day'})},
            {date: new Date(2019, 4, 30), labels: expect.objectContaining({en: 'Ascension Day'})},
            {date: new Date(2019, 5, 10), labels: expect.objectContaining({en: 'Whit Monday'})},
            {date: new Date(2019, 6, 21), labels: expect.objectContaining({en: 'National Holiday'})},
            {date: new Date(2019, 7, 15), labels: expect.objectContaining({en: 'Assumption'})},
            {date: new Date(2019, 10, 1), labels: expect.objectContaining({en: 'All Saints\' Day'})},
            {date: new Date(2019, 10, 11), labels: expect.objectContaining({en: 'Armistice'})},
            {date: new Date(2019, 11, 25), labels: expect.objectContaining({en: 'Christmas Day'})},
        ]);
    });
});

describe('Easter Monday', () => {
    test('2022', () => {
        expect(isHoliday(new Date(2022, 3, 18, 12, 10, 1))).toBe(true);
    });

    test('2020', () => {
        expect(isHoliday(new Date(2020, 3, 13))).toBe(true);
    });

    test('2018', () => {
        expect(isHoliday(new Date(2018, 3, 2))).toBe(true);
    });
});

describe('Ascension day', () => {
    test('2022', () => {
        expect(isHoliday(new Date(2022, 4, 26))).toBe(true);
    });

    test('2020', () => {
        expect(isHoliday(new Date(2020, 4, 21, 14, 2, 1))).toBe(true);
    });

    test('2018', () => {
        expect(isHoliday(new Date(2018, 4, 10))).toBe(true);
    });
});

describe('Whit Monday', () => {
    test('2022', () => {
        expect(isHoliday(new Date(2022, 5, 6))).toBe(true);
    });

    test('2020', () => {
        expect(isHoliday(new Date(2020, 5, 1))).toBe(true);
    });

    test('2018', () => {
        expect(isHoliday(new Date(2018, 4, 21, 12, 5, 1))).toBe(true);
    });
});

describe('edge cases', () => {
    test('returns false when it is not a holiday', () => {
        expect(isHoliday(new Date(2022, 0, 2))).toBeFalsy();
        expect(isHoliday(new Date(2022, 2, 6))).toBeFalsy();
    });

    test('returns false if no date was provided', () => {
        const noDate = undefined as unknown as Date;
        expect(isHoliday(noDate)).toBe(false);
    })
});


src/belgianHolidays.ts
function getEaster(year: number) {
    const C = Math.floor(year / 100);
    const N = year - 19 * Math.floor(year / 19);
    const K = Math.floor((C - 17) / 25);
    let I = C - Math.floor(C / 4) - Math.floor((C - K) / 3) + 19 * N + 15;
    I = I - 30 * Math.floor((I / 30));
    I = I - Math.floor(I / 28) * (1 - Math.floor(I / 28) * Math.floor(29 / (I + 1)) * Math.floor((21 - N) / 11));
    let J = year + Math.floor(year / 4) + I + 2 - C + Math.floor(C / 4);
    J = J - 7 * Math.floor(J / 7);
    const L = I - J;
    const M = 3 + Math.floor((L + 40) / 44);
    const D = L + 28 - 31 * Math.floor(M / 4);

    return new Date(year, M - 1, D);
}

function getEasterMonday(year: number) {
    const easter = getEaster(year);
    const daysToMonday = ((7 - easter.getDay()) % 7 + 1) % 7;
    return new Date(easter.getFullYear(), easter.getMonth(), easter.getDate() + daysToMonday);
}

function getAscensionDay(year: number) {
    let date = new Date(getEaster(year));
    date.setDate(date.getDate() + 39);
    return date;
}

function getWhitMonday(year: number) {
    let whitMonday = new Date(getAscensionDay(year));
    whitMonday.setDate(whitMonday.getDate() + 9);
    const daysToMonday = ((7 - whitMonday.getDay()) % 7 + 1) % 7;
    whitMonday.setDate(whitMonday.getDate() + daysToMonday);
    return whitMonday;
}


/** Get all official Belgian holidays in the provided year */
export function getHolidays(year: number): Holiday[] {
    return [
        {date: new Date(year, 0, 1), labels: {nl: 'Nieuwjaar', fr: 'Nouvel An', de: 'Neujahr', en: 'New Year\'s Day'}},
        {date: getEasterMonday(year), labels: {nl:'Tweede paasdag', fr: 'Lundi de Pâques', de: 'Ostermontag', en: 'Easter Monday'}},
        {date: new Date(year, 4, 1), labels: {nl:'Dag van de Arbeid', fr: 'Fête du travail', de: 'Tag der Arbeit', en: 'Labour day'}},
        {date: getAscensionDay(year), labels: {nl:'O.L.H. Hemelvaart', fr: 'Ascension', de: 'Christi Himmelfahrt', en: 'Ascension Day'}},
        {date: getWhitMonday(year), labels: {nl:'Tweede pinksterdag', fr: 'Lundi de Pentecôte', de: 'Pfingstmontag', en: 'Whit Monday'}},
        {date: new Date(year, 6, 21), labels: {nl: 'Nationale feestdag', fr: 'Fête nationale', de: 'Nationalfeiertag', en: 'National Holiday'}},
        {date: new Date(year, 7, 15), labels: {nl:'O.L.V. Hemelvaart', fr: 'Assomption', de: 'Mariä Himmelfahrt', en: 'Assumption'}},
        {date: new Date(year, 10, 1), labels: {nl:'Allerheiligen', fr: 'Toussaint', de: 'Allerheiligen', en: 'All Saints\' Day'}},
        {date: new Date(year, 10, 11), labels: {nl:'Wapenstilstand', fr:  'Armistice', de: 'Waffenstillstand', en: 'Armistice'}},
        {date: new Date(year, 11, 25), labels: {nl:'Kerstmis', fr: 'Noël', de: 'Weihnachten', en: 'Christmas Day'}}
    ];
}


/** Returns whether the provided Date is an official Belgian holiday */
export function isHoliday(date: Date): boolean {
    if (!date) {
        return false;
    }

    return getHolidays(date.getFullYear()).some(item => {
        return item.date.getDate() === date.getDate() &&
            item.date.getMonth() === date.getMonth() &&
            item.date.getFullYear() === date.getFullYear();
    });
}

export type Holiday = {
    date: Date;
    labels: {
        nl: string;
        fr: string;
        de: string;
        en: string;
    }
}
